### PR TITLE
Fixes machinery breaking when the area is edited/renamed via station blueprints & Deletes empty areas, ARCD upgrades

### LINKED ISSUE #70430: Area editing breaks air alarm  configuration
**Round ID** : N/A

**Reproduction**

1) Create a small area shown  and use area editor/small create area icon on the bottom right of your screen. Name the area to "OUTER" &  Install apc, air alarm , and 2 scrubbers just for testing like as shown  

Screenshot (26)

2) Create a second walled area within this area as shown. Use AreaEditor and name this area to "INNER"

Screenshot (27)

**3 Bugs Occur**

1) The air alarm installed in the "OUTER" Area also shows the scrubber installed in the "INNER" area. This shouldn't be possible as only the air alarm installed in the "INNER" area should locate this scrubber and the "OUTER" air alarm should no longer have control over that scrubber

2) If you unwrench & wrench back the "INNER" scrubber it breaks the air alarm interface located in "OUTER"

Screenshot (28)

Notice how the 2nd scrubber name isn't displayed and the off button is misaligned and i cant interact with this scrubber

3) If you install an APC & an Air alarm interface inside "INNER" but later break down a wall and merge area "INNER" to "OUTER" it breaks area "INNER" air alarm. i.e interface becomes non responsive

Screenshot (29)

Notice how the air alarm is still called "Inner air alarm" even though i merged this area with outer. The interface looks ok but you cant click on any buttons on the scrubber

### PR DESCRIPTION
**Pull request by SyncIt21**
## About The Pull Request
1. Fixes #70430
- APC's
- Air Alarms
- Fire Alarms
- Vents
- Scrubbers

Now all re-assign themselves to their correct areas & properly rename themselves after an area has been created, modified via the station blueprints.

2. Now Area's can't be expanded if both the current area where the player is standing and the selected area both have an apc because that would make the final area have 2 apc's

3. When 2 areas are merged, If the old area has no more turfs left then it is deleted 

4. Bonus fixes
- is removed redundant `Initialize()` proc inside light switch.
- arcd has all upgrades installed

## Changelog
:cl:
fix: area machinery from breaking after creating & modifying areas via the station blueprints.
refactor: turfs are cannonized from areas during creation/editing operations & empty areas are deleted
refactor: removed duplicate Initialize() from light switch
refactor: arcd has all upgrades installed
/:cl:

### CODE REVIEW

File: code/__HELPERS/areas.dm
Comment (ZephyrTFA) [inline code/__HELPERS/areas.dm:135]:
```suggestion
		creator.balloon_message(creator, "Conflicting APCs, remove one!")
```

Comment (san7890) [reply code/__HELPERS/areas.dm:135]: balloon alerts have to be all undercase (at least for `Conflicting`). might be better to do something like `too many APCs!`, that seems informative enough to me IMO. (acronyms being capitalized appear to be fine looking at the code, for examples like `TK` for telekinesis and `DNA`)

File: code/__HELPERS/areas.dm
Comment (LemonInTheDark) [inline code/__HELPERS/areas.dm:195]: this doesn't "save memory" it potentially dumps cpu usage into this proc.
uncontained turfs will slowly drop away, you are doing the job of `/datum/controller/subsystem/area_contents`. Don't.

Comment (SyncIt21) [reply code/__HELPERS/areas.dm:195]: True but that creates a problem for us. Since `turfs_to_uncontain` gets cleared over time the proc `has_contained_turfs()` won't be able to determine over time if we need to delete the area or not.
https://github.com/tgstation/tgstation/blob/e7a5cfb3a8d219b9a6ae6330f5931491194f1510/code/game/area/areas.dm#L250-#L252
It will always return true when `turfs_to_uncontain()` becomes 0 when it gets cleared by the sub system and thus our empty area will exist forever. 

Unless we program the subsystem to delete emty areas this is nessassry

Comment (SyncIt21) [reply code/__HELPERS/areas.dm:195]: This also creates problems for admins who want to jump to areas. When the area becomes empty it still gets listed and they get the error "No where to jump"

Comment (SyncIt21) [reply code/__HELPERS/areas.dm:195]: Never mind I should have read the code first you were right.  I'll keep the above comments just for fun.  But yup I will change it

File: code/__HELPERS/areas.dm
Comment (SyncIt21) [reply code/__HELPERS/areas.dm:135]: How about "too many conflicting APCS?" I think that makes both sides happy

Comment (SyncIt21) [reply code/__HELPERS/areas.dm:135]: Got a syntax error. its `balloon_alert()` not `balloon_message()`

File: code/__HELPERS/areas.dm
Comment (san7890) [inline code/__HELPERS/areas.dm:91]:
```suggestion
	var/error = ""
```

doing it as "null" isn't exactly useful since this is going to be mutated into a string anyways, so it should be typed as a string like such.

File: code/__HELPERS/areas.dm
Comment (san7890) [inline code/__HELPERS/areas.dm:98]:
```suggestion
	if(error)
```

empty strings are falsey (in conjunction with the earlier review about not typing it)

File: code/__HELPERS/areas.dm
Comment (san7890) [inline code/__HELPERS/areas.dm:131]:
```suggestion
		creator.balloon_alert(creator, "no area change!")
```

if you come up with something better sub that in instead

File: code/__HELPERS/areas.dm
Comment (san7890) [inline code/__HELPERS/areas.dm:133]:
```suggestion
		return

	//when expanding one area into another we don't want to merge their apcs. Display warning to user to dismantle an apc before proceeding
```

spacing nit

File: code/__HELPERS/areas.dm
Comment (san7890) [inline code/__HELPERS/areas.dm:194]:
```suggestion
			message_admins("Area [area_name] has been deleted!")
```

Does this need to be a message admins, anyways? If this is a failstate, would it be apt to have it be a stack trace instead?

File: code/game/objects/items/blueprints.dm
Comment (san7890) [inline code/game/objects/items/blueprints.dm:245]:
```suggestion
		/obj/machinery/light_switch,
```

also, alphabetize this list. there's a few vscode extensions that do this rather easily.

**Review by san7890 — CHANGES_REQUESTED** [review]
few nits

File: code/game/area/areas.dm
Comment (san7890) [inline code/game/area/areas.dm:291]: Why Cut() and then set it to null? Why not just null it out in the first place?

File: code/__HELPERS/areas.dm
Comment (LemonInTheDark) [reply code/__HELPERS/areas.dm:194]: why do we need to tell anyone?

Comment (SyncIt21) [reply code/__HELPERS/areas.dm:194]: > why do we need to tell anyone?

Fair enough. I'll remove it completely

File: code/game/area/areas.dm
Comment (SyncIt21) [reply code/game/area/areas.dm:291]: I guess i'm over cautious about memory leaks. I'll do that

**Comment by SyncIt21** [comment]
https://github.com/tgstation/tgstation/blob/b1a771dacb63237f56083306814514d3bf85c41c/code/game/machinery/_machinery.dm#L207-L208

@san7890 Don't know why they cut the list and then nullified it. But since it looks safer I'll do that too

**Comment by SyncIt21** [comment]
This seems more complicated than it looks. We have to look for areas for all turfs and not just oldA when reconfiguring airalarms etc. Need to think this more thoroughly.

File: code/__HELPERS/areas.dm
Comment (LemonInTheDark) [reply code/__HELPERS/areas.dm]: 👍

**Comment by LemonInTheDark** [comment]
you would cut and null if something else was holding a reference to the list and you wanted to clear what it was reading too. Shouldn't feel like you need to do it unless ya really want to

**Comment by SyncIt21** [comment]
> you would cut and null if something else was holding a reference to the list and you wanted to clear what it was reading too. Shouldn't feel like you need to do it unless ya really want to

Final decision. Should we to be extra safe or not?

**Comment by SyncIt21** [comment]
One fact we didn't account for is 2 or more areas are separated by walls and later if we break down those walls and merge all those areas into one then we have to reconfigure the machinery for all the merged areas

For example go to meta station atmospherics and create a rectangular area as such
Screenshot (126)

One could divide this area into 3 areas as follows
Screenshot (127)

Now say someone stands in the middle and breaks down 2 walls and they want to merge all 3 areas into 1 as follows
Screenshot (128)

Then all machinery from all 3 areas have to be reconfigured. Well now that's accounted for

File: code/__HELPERS/areas.dm
Comment (ZephyrTFA) [inline code/__HELPERS/areas.dm:113]:
```suggestion
		if(length(apc_map) > 1)
```

File: code/__HELPERS/areas.dm
Comment (ZephyrTFA) [inline code/__HELPERS/areas.dm:177]:
```suggestion
		for(var/obj/machinery/door/firedoor/firedoor as anything in merged_area.firedoors)
			firedoor.CalculateAffectingAreas()
```

File: code/__HELPERS/areas.dm
Comment (ZephyrTFA) [inline code/__HELPERS/areas.dm:181]: can this be `as anything` ?

File: code/__HELPERS/areas.dm
Comment (ZephyrTFA) [inline code/__HELPERS/areas.dm:186]: and this

File: code/__HELPERS/areas.dm
Comment (ZephyrTFA) [inline code/__HELPERS/areas.dm:191]: this too

File: code/__HELPERS/areas.dm
Comment (ZephyrTFA) [inline code/__HELPERS/areas.dm:196]: ditto

File: code/__HELPERS/areas.dm
Comment (ZephyrTFA) [inline code/__HELPERS/areas.dm:215]: I imagine all of the for loops in this can be `as anything`

**Review by ZephyrTFA — CHANGES_REQUESTED** [review]

**Comment by SyncIt21** [comment]
COMSIG_AREA_CREATED is now being modified to accept a list of old areas overwritten as explained in my comment. 
https://github.com/tgstation/tgstation/pull/73850#issuecomment-1462122670

**Comment by SyncIt21** [comment]
Screenshot (130)
caused by `disconnect_area_machinery()` returning not specifically type checking the stuff in its `as anything` for loop. time to revert

**Comment by SyncIt21** [comment]
Now thinking about it there are mapping subtypes of airalarm, firealarm etc we don't want to miss. Yup `as anything` is necessary

**Comment by SyncIt21** [comment]
I have over engineered this. Sending ` COSMIG_ENTER_AREA` and the `COSMIG_EXIT_AREA` to each atom of the affected turf is more than sufficient. Had to override the on_enter_area() and the on_exit_area() of specific machinery with the exception of apc to fully make this work & thats it

File: code/modules/power/apc/apc_main.dm
Comment (ZephyrTFA) [inline code/modules/power/apc/apc_main.dm:228]: convert to early return

File: code/modules/power/apc/apc_main.dm
Comment (ZephyrTFA) [inline code/modules/power/apc/apc_main.dm:205]:
```suggestion
	return ..()
```

File: code/modules/atmospherics/machinery/components/unary_devices/vent_scrubber.dm
Comment (ZephyrTFA) [inline code/modules/atmospherics/machinery/components/unary_devices/vent_scrubber.dm:72]: This shouldn't silently pass and should stack_trace if the passed in area is null

**Comment by SyncIt21** [comment]
Screenshot (131)
Looks like null is expected sometimes

**Comment by SyncIt21** [comment]
Merge, Review Anyone? I have no pending work on this.

**Review by ZephyrTFA — APPROVED** [review]

**Comment by ZephyrTFA** [comment]
`I don't get it but whatever` as a commit name does not instill confidence

**Comment by SyncIt21** [comment]
> `I don't get it but whatever` as a commit name does not instill confidence

I'm being honest though. Why it fails i don't understand, probably should ask lemon but this works just as well

**Comment by SyncIt21** [comment]
Can someone please merge this before we have another merge conflict? I have tested this extensively.

**Comment by ZephyrTFA** [comment]
I'm gonna stick my neck out for you here; if it breaks Lemon will break my kneecaps

**Pull request merged**
